katportalclient/client.py

```diff
@@ -1254,73 +1195,40 @@ def sensor_history(self, sensor_name, start_time_sec, end_time_sec,
             - If the request timed out
         """
-        # create new namespace and state variables per query, to allow multiple
-        # request simultaneously
-        state = {
-            'sensor': sensor_name,
-            'done_event': tornado.locks.Event(),
-            'num_samples_pending': 0,
-            'include_value_ts': include_value_ts,
-            'samples': []
-        }
-        namespace = str(uuid.uuid4())
-        self._sensor_history_states[namespace] = state
-        # ensure connected, and subscribed before sending request
-        yield self.connect()
-        yield self.subscribe(namespace, ['*'])
-
         params = {
             'sensor': sensor_name,
-            'time_type': SAMPLE_HISTORY_REQUEST_TIME_TYPE,
-            'start': start_time_sec * SAMPLE_HISTORY_REQUEST_MULTIPLIER_TO_SEC,
-            'end': end_time_sec * SAMPLE_HISTORY_REQUEST_MULTIPLIER_TO_SEC,
-            'namespace': namespace,
-            'request_in_chunks': 1,
-            'chunk_size': SAMPLE_HISTORY_CHUNK_SIZE,
-            'limit': MAX_SAMPLES_PER_HISTORY_QUERY
+            'start_time': start_time_sec,
+            'end_time': end_time_sec,
+            'limit': MAX_SAMPLES_PER_HISTORY_QUERY,
+            'all_fields': True,
+            'timeout': timeout_sec
         }
         url = url_concat(
-            self.sitemap['historic_sensor_values'] + '/samples', params)
+            self.sitemap['historic_sensor_values'] + '/query', params)
         self._logger.debug("Sensor history request: %s", url)
         response = yield self._http_client.fetch(url)
         data = json.loads(response.body)
-        if isinstance(data, dict) and data['result'] == 'success':
-            download_start_sec = time.time()
-            # Query accepted by portal - data will be returned via websocket, but
-            # we need to wait until it has arrived.  For synchronisation, we wait
-            # for a 'done_event'. This event is updated in
-            # _process_redis_message().
-            try:
-                timeout_delta = timedelta(seconds=timeout_sec)
-                yield state['done_event'].wait(timeout=timeout_delta)
-
-                self._logger.debug('Done in %d seconds, fetched %s samples.' % (
-                    time.time() - download_start_sec,
-                    len(state['samples'])))
-            except tornado.gen.TimeoutError:
-                raise SensorHistoryRequestError(
-                    "Sensor history request timed out")
-
-        else:
+        if 'data' not in data:
             raise SensorHistoryRequestError("Error requesting sensor history: {}"
                                             .format(response.body))
 
-        def sort_by_timestamp(sample):
-            return sample.timestamp
+        samples = []
+        for sample in data['data']:
+            if include_value_ts:
+                # Requesting value_timestamp in addition to
```

Comment:
It seems like very heavy lifting to reprocess each sample from a result loaded as json, just to select fields and then put it back into json. SQL can do this much better.
Think about whether you cannot ensure that the json.load(result) is usable as it is returned, such that you need to only do the sort here and does not have to reprocess each sample. Is this possible?
May be by extending the query api with a parameter to select the specific list of fields you need in the result (similar to how all_fields are implemented)?

Reply:
@lvdheever I don't quite follow this comment.  We were not putting it back into a JSON struct.  We were created named tuples here.  Which is an efficient way to store it.  We should keep those, as they were part of the API.